ygot/struct_validation_map.go

```diff
@@ -439,44 +450,105 @@ func copyInterfaceField(dstField, srcField reflect.Value) error {
+	// Skip cases where there are empty maps in both src and dst.
+	if srcField.Len() == 0 && dstField.Len() == 0 {
+		return nil
 	}
 
-	keys := srcField.MapKeys()
-	if k := srcField.MapIndex(keys[0]).Kind(); k != reflect.Ptr {
-		return fmt.Errorf("invalid map, got member type %s", k)
+	m, err := validateMap(srcField, dstField)
+	if err != nil {
+		return err
 	}
 
+	srcKeys := srcField.MapKeys()
+	dstKeys := dstField.MapKeys()
+
 	// TODO(robjs): When we move to go1.9+ only support in ygot, we can use the
 	// following to make a map of a particular size.
 	// nm := reflect.MakeMapWithSize(reflect.MapOf((keys[0]).Type(), srcField.MapIndex(keys[0]).Type()), srcField.Len())
-	nm := reflect.MakeMap(reflect.MapOf((keys[0]).Type(), srcField.MapIndex(keys[0]).Type()))
-	for _, k := range keys {
-		v := srcField.MapIndex(k)
-		d := reflect.New(v.Elem().Type())
-		if err := copyStruct(d.Elem(), v.Elem()); err != nil {
-			return err
+	nm := reflect.MakeMap(reflect.MapOf(m.key, m.value))
+
+	mapsToMap := []struct {
+		keys  []reflect.Value
+		field reflect.Value
+	}{
+		{srcKeys, srcField},
+		{dstKeys, dstField},
+	}
+	existingKeys := map[interface{}]bool{}
+
+	for _, m := range mapsToMap {
+		for _, k := range m.keys {
+			// Check that this key has not already been mapped. We do not support
+			// the case where there are overlapping keys.
+			// TODO(robjs): Implement mapping of entities that exist in both maps.
+			if _, ok := existingKeys[k.Interface()]; ok {
+				return fmt.Errorf("cannot map element %v, overlaps in both maps", k.Interface())
+			}
+			existingKeys[k.Interface()] = true
+
+			v := m.field.MapIndex(k)
+			d := reflect.New(v.Elem().Type())
+			if err := copyStruct(d.Elem(), v.Elem()); err != nil {
+				return err
+			}
+			nm.SetMapIndex(k, d)
 		}
-		nm.SetMapIndex(k, d)
 	}
 	dstField.Set(nm)
 	return nil
 }
 
+// mapTypes provides a specification of a map.
+type mapType struct {
+	key   reflect.Type // key is the type of the key of the map.
+	value reflect.Type // value is the type of the value of the map.
+}
+
+// validateMap checks the srcField and dstField reflect.Value structs
+// to ensure that they are valid maps of struct pointers, and that their keys
+// types are the same. It returns a specification of the map type if the maps
+// match.
+func validateMap(srcField, dstField reflect.Value) (*mapType, error) {
+	if s := srcField.Kind(); s != reflect.Map {
+		return nil, fmt.Errorf("invalid src field, was not a map, was: %v", s)
+	}
+
+	if d := dstField.Kind(); d != reflect.Map {
+		return nil, fmt.Errorf("invalid dst field, was not a map, was: %v", d)
+	}
+
+	st, dt := srcField.Type(), dstField.Type()
+	se, de := st.Elem(), dt.Elem()
+	if se != de {
+		return nil, fmt.Errorf("invalid maps, src and dst value types are different, %v != %v", se, de)
+	}
+
+	if !util.IsTypeStructPtr(se) || !util.IsTypeStructPtr(de) {
```

Comment:
A list member must be a struct in the output code -- i.e.,:

```yang
list foo {
  key "bar";
  leaf bar { type string; }
}
```

So even if we have a `leaf-list` there, then we need to have a `struct` that encapsulates the `slice` - so no way this can be `map[<some type>]<some slice>`.